interop-testing/src/test/java/io/grpc/testing/integration/CascadingTest.java

```diff
@@ -0,0 +1,316 @@
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.testing.integration;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.fail;
+
+import io.grpc.Context;
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor;
+import io.grpc.ServerCall;
+import io.grpc.ServerCallHandler;
+import io.grpc.ServerInterceptor;
+import io.grpc.ServerInterceptors;
+import io.grpc.Status;
+import io.grpc.StatusRuntimeException;
+import io.grpc.inprocess.InProcessChannelBuilder;
+import io.grpc.inprocess.InProcessServerBuilder;
+import io.grpc.internal.ManagedChannelImpl;
+import io.grpc.internal.ServerImpl;
+import io.grpc.stub.StreamObserver;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+
+import java.io.IOException;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Integration test for various forms of cancellation & deadline propagation.
+ */
+@RunWith(JUnit4.class)
+public class CascadingTest {
+
+  @Mock
+  TestServiceGrpc.TestService service;
+  private ManagedChannelImpl channel;
+  private ServerImpl server;
+  private AtomicInteger depth;
+  private AtomicInteger observedCancellations;
+  private AtomicInteger receivedCancellations;
+  private TestServiceGrpc.TestServiceBlockingStub blockingStub;
+  private TestServiceGrpc.TestServiceStub asyncStub;
+  private ScheduledExecutorService scheduler;
+
+  @Before
+  public void setUp() throws Exception {
+    MockitoAnnotations.initMocks(this);
+    channel = InProcessChannelBuilder.forName("channel").build();
+    depth = new AtomicInteger();
+    observedCancellations = new AtomicInteger();
+    receivedCancellations = new AtomicInteger();
+    blockingStub = TestServiceGrpc.newBlockingStub(channel);
+    asyncStub = TestServiceGrpc.newStub(channel);
+    scheduler = Executors.newScheduledThreadPool(1);
+  }
+
+  @After
+  public void tearDown() {
+    Context.ROOT.attach();
+    channel.shutdownNow();
+    server.shutdownNow();
+  }
+
+  /**
+   * Test {@link Context} expiration propagates from the first node in the call chain all the way
+   * to the last.
+   */
+  @Test
+  public void testCascadingCancellationViaOuterContextExpiration() throws Exception {
+    startChainingServer(3);
+    Context.current().withDeadlineAfter(150, TimeUnit.MILLISECONDS, scheduler).attach();
+    try {
+      blockingStub.unaryCall(Messages.SimpleRequest.getDefaultInstance());
+      fail("Expected cancellation");
+    } catch (StatusRuntimeException sre) {
+      // Wait for the workers to finish
+      Thread.sleep(500);
```

Comment:
Instead of sleeping, maybe shutdown the server and the server's executor (we'd need to provide one) and then do `awaitTermination()` on each? A bit painful but should be reliable and much faster than waiting half a second.